optaplanner-wb-screens/optaplanner-wb-solver-editor/optaplanner-wb-solver-editor-client/src/main/java/org/optaplanner/workbench/screens/solver/client/editor/TerminationConfigForm.java

```diff
@@ -41,73 +35,69 @@
-            TerminationTreeItemContent terminationTreeItemContent = new TerminationTreeItemContent( TerminationConfigForm.this, nestedTreeItem, terminationConfigOption, terminationConfigModel, value.getK1() );
-            nestedTreeItem.setUserObject( terminationTreeItemContent );
-            nestedTreeItem.setWidget( terminationTreeItemContent );
+        treeItemContent.removeDropDownOption( TerminationConfigOption.valueOf( terminationType ) );
+        view.displayEmptyTreeLabel( rootTreeItem.getChildCount() == 0 );
+        view.refreshTree();
+    }
 
-            terminationPopup.resetSelectedTermination();
-            terminationPopup.setErrorMessage( null );
-            if ( creationFinished ) {
-                terminationPopup.hide();
-            } else {
-                terminationPopup.init( treeItemContent, getAssignableTerminationConfigOptions( treeItemContent.getModel() ) );
+    private void insertTreeItem( TreeItem treeItem, TreeItem nestedTreeItem, TerminationConfigOption terminationConfigOption ) {
+        if ( treeItem.getChildCount() == 0 ) {
+            treeItem.addItem( nestedTreeItem );
+        } else {
+            for ( int i = 0; i < treeItem.getChildCount(); i++ ) {
+                TerminationTreeItemContent childTreeItemContent = (TerminationTreeItemContent) treeItem.getChild( i ).getUserObject();
+                if ( terminationConfigOption.ordinal() < childTreeItemContent.getTerminationConfigOption().ordinal() ) {
+                    treeItem.insertItem( i, nestedTreeItem );
+                    break;
+                }
+                if ( i == treeItem.getChildCount() - 1 ) {
+                    treeItem.addItem( nestedTreeItem );
+                    break;
+                }
             }
         }
     }
 
-    public List<Pair<String, String>> getAssignableTerminationConfigOptions( TerminationConfigModel terminationConfigModel ) {
-        List<Pair<String, String>> uninitializedTerminationConfigOptions = new ArrayList<>();
-        for ( TerminationConfigOption terminationConfigOption : TerminationConfigOption.values() ) {
-            Pair<String, Object> terminationValue = getModelManager( terminationConfigOption ).getTerminationValue( terminationConfigModel );
-            if ( terminationValue.getK2() == null || terminationConfigOption.equals( TerminationConfigOption.NESTED ) ) {
-                uninitializedTerminationConfigOptions.add( new Pair<>( getModelManager( terminationConfigOption ).getLocalizedTerminationConfigOption(), terminationConfigOption.name() ) );
-            }
-        }
-        return uninitializedTerminationConfigOptions;
+    private TerminationTreeItemContent createTerminationTreeItemContent( TerminationConfigOption terminationConfigOption, TreeItem treeItem, TerminationConfigModel terminationConfigModel ) {
+        TerminationTreeItemContent terminationTreeItemContent = IOC.getBeanManager().lookupBean( TerminationTreeItemContent.class ).newInstance();
```

Comment:
You should probably store the reference to ```terminationTreeItemContent``` (all instances you may create) and destroy them with ```IOC.getBeanManager().destroy( o )``` otherwise you risk a memory leak.. IIRC @ederign @mbarkley have an annotation that may make garbage collection of manually instantiated beans easier too.. but you'd need to ask them about it's use.